Http3

constanthatz opened the pull request.

constanthatz committed:
change reponse functions to build repsonses from lists and to use proper crlf.

constanthatz committed:
add while loop to capture entire client message.

constanthatz committed:
new error code generating function.

constanthatz committed:
with server socket function and parse request tests.

constanthatz committed:
add example web home directory.

constanthatz committed:
add response uri function and supporting function.

constanthatz committed:
fix resolve uri function to deal with errors.

constanthatz committed:
generalize error and ok reponse, with appropriate changes to gen_list and gen_text.

constanthatz committed:
add tests

constanthatz committed:
initialize HTTP server script.

constanthatz committed:
rename files

constanthatz committed:
initialize HTTP concurrent server.

constanthatz committed:
remove test code.

constanthatz committed:
add gevent code block.

constanthatz committed:
first pass at incorporating gevent.

constanthatz committed:
add gevent stream server.

constanthatz committed:
clean up tests.

constanthatz committed:
revert tests.

constanthatz committed:
fix socket close

cewing commented on HTTP_server.py:
```diff
@@ -0,0 +1,128 @@
+    first_line = 'HTTP/1.1 200 OK'
+    timestamp = 'Date: ' + email.utils.formatdate(usegmt=True)
+    content_header = 'Content-Type: {}'.format(args[0][0])
+    body = '''<!DOCTYPE html>\n<html lang="en">\n<head>\n<meta charset="utf-8">\n</head>\n<body>\n{}</body>\n</html>'''.format(args[0][1])
```
The entire purpose of triple-quoted strings is to allow you to break lines.  You can make this look a lot neater by doing something like this:

``` python

HTML_PAGE = """
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
  </head>
  <body>
{}
  </body>
</html>
"""

def response_ok(*args):
    # ...
    body = HTML_PAGE.format(args[0][1])
    #...
```

You can even share the `HTML_PAGE` constant you've defined with the `response_error` function below, since it's the same thing.  

You get pep8 wins, DRY code wins and easier code to read all in one.

cewing commented on HTTP_server.py:
```diff
@@ -0,0 +1,128 @@
+        info = (content_type, body)
+    elif os.path.isfile(uri):
+        body = gen_text(uri)
+        content_type = 'text/html'
```
This is not right.  Not every item you view will be an html page.  figure out how to get the right content-type value

cewing commented on HTTP_server.py:
```diff
@@ -0,0 +1,128 @@
+                    info = resolve_uri(uri)
+                    response = response_ok(info)
+                except RequestError, msg:
+                    errors = str(msg).strip("'").split(' ', 1)
```
I know we sat and fixed this up today.  Where are those fixes?

cewing commented on http_concurrent_server.py:
```diff
@@ -0,0 +1,133 @@
+
+            # conn, addr = server_socket.accept()
+            try:
+                while True:
```
everything from this line up (except the buffer size assignment) should be removed.  Again, we fixed this today.  Where are those fixes?

cewing commented on http_concurrent_server.py:
```diff
@@ -0,0 +1,133 @@
+    except KeyboardInterrupt:
+        socket.close()
+
+
```
Don't repeat all this code, import it.

constanthatz committed:
change something in every function.

http_server:
fix __str__ in  RequestError
change to gen_Text to use with and rb to read as binary and handle closing better
set home directory to cwd and fix character stripping from input in resolve_uri.

constanthatz committed:
fix all tests to match new server code.

constanthatz committed:
add content length header.
